[STEP 1] JusBug, maxhyunm

maxhyunm opened the pull request:
@gzzjk159
안녕하세요! 커리어스타터 9기 6조의 JusBug, maxhyunm 입니다.
숫자야구 STEP 1 PR 올려드립니다!
잘 부탁드립니다~

## 고민했던 사항, 비고사항
### 순서도
프로젝트 내용을 파악하는 시간을 가질겸 각자가 생각하는 순서도를 먼저 그려보았고, 해당 내용을 통합하여 하나의 순서도로 정리하였습니다.
옵셔널 !의 사용제약과 3중 탭 사용 제약 등의 조건을 생각하며 작업하다보니 반복/조건문 내용 관련하여 고민이 많았지만 고차함수 사용을 통해 해결하였습니다.
- JusBug : 처음에는 순서도를 작성할 때 단순히 각각의 파라미터와 함수들을 구체적으로 정의하지 않고 작성했는데 그러다 보니 맥스와 함께 알고리즘을 구성해가는 과정에서 이해하기 부족한 상황이 발생했었습니다. 이를 개선하기 위해 함수 종류에 따라 순서도를 효율적으로 배치하고 각자의 함수 안에도 구체적인 변수와 연산을 넣어 바로 넣어 보다 효율적인 소통과 코딩으로 이어질 수 있도록 노력했습니다.
- maxhyunm : 순서도를 최대한 상세하게 정리하는 것에 집중하였습니다. 흐름의 분기점이나 반복, 특별한 행위에 대한 설명을 모두 정리하여 작업의 뼈대로서의 역할에는 충실했으나 설명문이 너무 길어져 다소 보기 어려운 모양이 되었습니다. 이 부분은 JusBug가 코드 형식으로 다시 정리를 해주셔서 불필요한 내용을 많이 쳐낼 수 있었습니다.

### 짝프로그래밍
- JusBug : 평소에는 혼자 조용한 환경에서 코딩하는 습관이 들어서 그런지 네비게이터로서 타입과 이름들을 바로 구상해서 제 생각을 맥스에게 말로 정확하게 전달하는 것이 굉장히 어색하고 힘들었습니다. 오히려 말하면서 잘 못되지 않았을까 하는 두려움이 컸던 것 같습니다. 짝프로그래밍을 통해서 다시 한번 많이 부족하다는 것을 스스로 느꼈고 팀에게 도움이 될 수 있도록 코딩 습관의 개선이 필요함을 느꼈습니다.
- maxhyunm : 순서도에 일부 정리가 되어 있다고는 해도, 머릿속에만 있던 내용을 말로 설명하는 것이 굉장히 어려웠습니다. 또 내비게이터가 설명을 주실 때 올바르게 받아적기도 익숙지 않아 생각보다 많은 시간을 소요하게 된 것 같습니다. 상대방이 내비게이터를 맡았을 때 질문이나 의견을 제시할 수 없는 점을 자꾸 잊어버리게 되어 반성했습니다...! 또 커밋 메시지 내용을 먼저 통일하지 못했던 점이 조금 아쉽습니다.

maxhyunm committed:
feat 랜덤 배열 함수 작성(1)

내비게이터 : JusBug
드라이버 : Max

JusBug committed:
feat 숫자 비교 함수

네비게이터 : 맥스
드라이버 : jusbug

maxhyunm committed:
feat 게임시작 함수 생성

내비게이터 : JusBug
드라이버 : 맥스#

JusBug committed:
feat 메뉴 시작 함수

네비게이터 : 맥스
드라이버 : jusbug

maxhyunm committed:
feat: STEP 1 기본 기능 구현 완료

내비게이터 : JusBug
드라이버 : Max

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
 import Foundation
 
-print("Hello, World!")
+typealias resultType = (strike: Int, ball: Int)
```
typealias를 적용하여 코드를 작성하신 것은 정말 좋으신 것 같습니다. 
하지만 Result Type은 swift에 존재합니다. 다른 이름을 적용하는 것이 좋아보이네요.
[API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/) 네이밍 참고해주세요

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
 
```
```swift
//
//  NumberBaseball - main.swift
//  Created by yagom. 
//  Copyright © yagom academy. All rights reserved.
// 
```
현재는 주석이 yagom으로 되어있네요. 보통 주석내용은  파일이름, 작성자, 날짜(작성 날짜만)가 들어가는 편입니다. 나중에 편하실 때 수정해주시면 좋을 거 같습니다.

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
-print("Hello, World!")
+typealias resultType = (strike: Int, ball: Int)
+
+func randomNumbers() -> Array<Int> {
```
보통 배열을 표현할때 명사에 s를 붙여줍니다. `items = [item]` 이렇게 사용합니다. 따라서 이 코드는 배열 변수로 보일거 같은 네임인거 같습니다.  [API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/) 네이밍 참고하고 수정해주시면 좋을거 같습니다.

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+        if computerNumbers.count == 3 {
+            break
+        }
+        let randomNumber = Int.random(in:1...9)
```
현재 while문에서 조건을 달아줘서 아래 if문은 사용되지 않을 거 같은데 어떻게 생각하시나요?

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+            break
+        }
+        let randomNumber = Int.random(in:1...9)
+        if computerNumbers.contains(randomNumber) {
```
if문 말고 다른 조건문인 guard가 있습니다. 두 조건문의 차이점은 무엇일까요? (어떨 때 쓰는지만 알고 가면 좋을 것 같네요. 수정은 안하셔도 됩니다 😀)

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+typealias resultType = (strike: Int, ball: Int)
+
+func randomNumbers() -> Array<Int> {
+    var computerNumbers: Array<Int> = []
```
computerNumbers라는 네이밍은 컴퓨터만 사용할 거 같은 네이밍인데 아래의 코드를 보면 사용자도 사용을 하는 것으로 보입니다. 네이밍을 어떻게 하는 것이 좋을까요?

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+    return computerNumbers
+}
+
+func compareNumbers(user userNumbers: Array<Int>, to computerNumbers: Array<Int>) -> resultType {
```
전달인자를 적절하게 사용하셨네요. 👍

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+}
+
+func compareNumbers(user userNumbers: Array<Int>, to computerNumbers: Array<Int>) -> resultType {
+    let strikes = userNumbers.enumerated().filter {userNumbers[$0.offset] == computerNumbers[$0.offset]}.map {$0.element}
```
enumeraterd와 고차함수를 사용하여 strike의 개수를 찾으신 것은 정말 좋으신 거 같습니다. 
하지만 가독성이 좀 떨어져 보이네요.  고차함수를 연결할 때 아래와 같이 중괄호 사이를 띄는 컨벤션을 사용하는 편입니다. 
```suggestion
    let strikes = userNumbers.enumerated().filter { userNumbers[$0.offset] == computerNumbers[$0.offset] }.map { $0.element }
```
현재 코드가 100자를 넘어가는 것으로 보이는데 보통 100자를 넘기지 않게 코드를 작성하는 편입니다. 개행을 하거나 변수를 주어 코드수를 조금 줄여보시는 것이 어떨까요? 
[Style Guide](https://google.github.io/swift/)를 참고하시면 좋을거 같네요.
(팁! Xcode -> 설정 -> Text Editing -> Page guide at column 에서 가이드라인을 설정할 수 있어요!

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+
+func compareNumbers(user userNumbers: Array<Int>, to computerNumbers: Array<Int>) -> resultType {
+    let strikes = userNumbers.enumerated().filter {userNumbers[$0.offset] == computerNumbers[$0.offset]}.map {$0.element}
+    let balls = userNumbers.filter {(strikes.contains($0) == false) && (computerNumbers.contains($0))}
```
요것도 마찬가지입니다. 중괄호 컨벤션을 사용해보시는 것이 좋을거 같네요

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+}
+
+func menuSelect() {
+    var computerNumbers: Array<Int>
```
```suggestion
    let computerNumbers: Array<Int> = randomNumbers()
```
위와 같이 let을 사용할 수도 있습니다.  
var를 사용하신 이유가 있을까요?

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+    var chance: Int = 9
+    computerNumbers = randomNumbers()
+    while chance > 0 {
+        let userNumbers = randomNumbers()
```
let 사용하신 점이 정말 좋네요 👍

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+    computerNumbers = randomNumbers()
+    while chance > 0 {
+        let userNumbers = randomNumbers()
+        var result: resultType = compareNumbers(user: userNumbers, to: computerNumbers)
+        var printNumber = userNumbers.map {String($0)}.joined(separator: ", ")
```
고차함수의 활용은 좋아보이네요 LGTM👍

다만 var를 사용하신 이유가 있나요? let으로 선언해도 문제 없어 보이네요

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
 
 import Foundation
```
`import Foundation` 어떤 기능을 import하는 것일까요?
꼭 필요할까요?

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+        chance -= 1
+        if chance == 0 {
+            print("컴퓨터 승리...!")
+            break
```
```swift
while chance > 0 {
        let userNumbers = randomNumbers()
        var result: resultType = compareNumbers(user: userNumbers, to: computerNumbers)
        var printNumber = userNumbers.map {String($0)}.joined(separator: ", ")
        print("임의의 수 : \(printNumber)")
        print("\(result.strike) 스트라이크, \(result.ball) 볼")
        chance -= 1
        if chance == 0 {
            print("컴퓨터 승리...!")
            break
        } else if result.strike == 3 {
            print("사용자 승리!")
            break
        }
        print("남은 기회 : \(chance)")
    }
```
전체적으로 개행이 너무 안되어 있어서 가독성이 떨어지는 것 같습니다.
두분의 컨벤션을 정해서 띄어쓰기를 주시면 될거 같습니다.
[Style Guide](https://google.github.io/swift/) 참고용 가이드로 정답은 아니니 참고만 해주세요

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+    let strike = strikes.count
+    let ball = balls.count
+    
+    return (strike: strike, ball: ball)
```
typealias를 사용하여 튜플로 반환하신 점 정말 좋은거 같네요 LGTM 👍

gzzjk159 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,53 @@
+        }
+        computerNumbers.append(randomNumber)
+    }
+    return computerNumbers
```
Set이라는 컬렉션 타입도 있는데 한번 고려해서 작성해보시는것도 좋아보이네요
현재 고려사항은 아니니 다음 step에서 작성해도 좋을 거 같습니다.

gzzjk159 reviewed (changes requested):
안녕하세요 JusBug, maxhyunm 
저는 vetto(베토, 베또 편하신대로 불러주세요)입니다. 한 주동안 잘 부탁드립니다.

코드리뷰는 처음이라 이상한 말을 해도 조금의 양해 부탁드립니다.

코드의 정답은 없다고 생각합니다. 같은 문제를 줘도 전부 다 다르게 짤 수 있다고 생각합니다. 따라서 제가 조언을 해도 무조건 고치려고 하기보단 적절한 예시를 들면서 이해를 주시는 게 더 좋을 수 있습니다. 저도 코드를 보면서 모르는 것을 알아가는 느낌이라 좋은 거 같네요 😀

## 고민하신 사항
### 순서도 
많이 고민 하시고 적절히 그리신 거 같아서 정말 놀랐습니다. 고민하신 부분은 저도 처음에 작성할 때 충분히 고민했던 사안이네요. 너무 상세하게 작성하면 이해하기 번거로워지고, 간결하게 작성하면 이해력이 떨어지는 딜레마가 있습니다. 하지만 순서도에도 코드처럼 정답이 존재하지 않기 때문에, 두 명이 함께 서로의 생각을 조합하고 수정하면서 서로 이해할 수 있고, 다른 사람들도 쉽게 이해할 수 있는 코드나 순서도를 만드는 것이 가장 적절하다고 생각합니다.

### 짝프로그래밍
저도 짝 프로그래밍을 처음 접했을 때는 생소하고 어색한 경험이 있었습니다. 특히, 네비게이터는 머릿속으로만 생각하며 코드를 작성하고, 드라이버는 그대로 적어야 하는 점이 힘들었었습니다. 처음에는 규칙을 엄격하게 지키려고 했지만, 나중에는 변수와 함수명은 서로 상의해서 작성하고, 15분마다 교대하는 등 규칙을 유연하게 조정하면서 짝 프로그래밍을 진행하고 있습니다. 이렇게 형식에 너무 얽매이는 것보다는 두 명이 함께 상의하며 적절한 방법을 찾아가는 것이 더 좋다고 생각합니다.


## 칭찬 드리고 싶은 점
정말 빨리 step1을 수행하셨네요. 첫 주에는 적응 기간이 있을텐데 대단하신거 같습니다. 👍
제약사항 부분도 잘 지키신 점 정말 좋네요 👍
깔끔하게 함수 3개로 나누어서 작성하신 점 좋네요 👍

### 순서도
프로젝트를 참여하지 않은 제 3자가 봐도 이해하기 쉬운 순서도인것 같습니다. 이와 같이 순서도를 사용하게 되면 다른 사람의 이해도를 높일 수 있습니 다른 장점도 많으니 한번 찾아보시면 좋을 거 같습니다.

### 고차함수 사용
map,filter같은 고차함수를 잘 활용하셨네요. 고차함수를 잘 사용하면 길었던 코드를 확 줄일 수 있습니다. 나중에 다른 고차함수도 한번 공부해보시는 것을 추천 드립니다.

### 커밋 메세지
feat와 같이 커밋 메세지를 분류하신 것 정말 좋은거 같습니다. 다른 커멧 메세지는 사용하지 않으신거 같은데 어떤 것들이 있을까요? 다른 커밋메세지도 한번 찾아보시고 적용해보시는 것도 좋을거 같네요.


## 조언
### 요구사항
현재 요구사항에서는 전역변수를 생성하여 사용하는 것으로 되어있는데 코드에서는 메서드 안에 지역변수로 선언되어 전역변수로는 사용되지 않은 것 같네요. 요구사항을 지키면서 코드를 작성하는 것은 매우 중요하다고 생각합니다. 저같은 경우에도 요구사항을 매번 놓치는 것이 있습니다. 그래서 요구사항을 놓치는 것이 있는지 없는지 동기들한테 코드 조언을 구하면서 서로 도움을 얻는 방법을 사용하곤 합니다.

### 커밋 메세지
커밋 메세지를 보면 두분이 규칙을 정하고 남기고 있다는 것이 느껴지네요. 하지만 조금 짥은 명사만 가지고 작성되고 있어 정확하게 어떤 커밋인지 알 수 없을거 같네요. 조금에 디테일을 추가해보는 것이 어떨까요?
저도 완벽한건 아니지만 아래와 같이 쓰면서 어떤 메서드나 변수를 선언 했는지는 알려주고 있어요.

`feat: compareNumbers 메서드 구현`

커밋 메세지는 왜 남겨야 하는지, 어떻게 작성하는 것이 좋은 커밋 메세지인지 한번 고민해보는 것도 좋을거 같습니다.

### 커밋 브랜치
보통 step을 진행하게 되면 main브랜치에서 step브랜치를 따로 파서 거기서 코드를 진행하는 편입니다. 현재는 main브랜치에서 PR보내셨는데 다음부터는 step별로 나누어서 PR을 보내시는 것이 나중에 코드를 볼 때 좋으실 것 같습니다. 😀

maxhyunm committed:
feat: STEP 1 수정

- 변수명, 함수명 변경
- 개행 등 컨벤션 수정
- 불필요한 if문 제거
- 변경되지 않는 변수를 상수로 수정
- 요구사항에서 벗어나는 내용 수정(전역변수)

maxhyunm commented:
@gzzjk159 베토 안녕하세요! 상세한 리뷰 정말 감사드립니다...!🙇
베토께서 말씀주신 사항들을 기반으로 수정과 답변을 정리해 보았습니다.

## 수정한 사항
- resultType -> gameResultType 로 변경하였습니다.
- computerNumbers -> opponentNumbers 로 변경하여 해당 배열의 역할을 더 명확히 표현했습니다.
- randomNumbers -> makeRandomNumbers 로 동사를 붙여 함수명을 변경하였습니다.
- 개행을 추가하여 코드의 가독성을 높였습니다.
- opponentNumbers와 chance를 전역변수로 변경했습니다.
- 불필요한 if문을 제거하였습니다.
- var -> let으로 일부 변경하였습니다.

## 추가 답변
- guard와 if의 차이 : if는 해당 조건인 경우 중괄호 내의 코드를 실행하는 조건문이고, guard는 해당 조건에 부합하는 경우만 하단의 코드로 넘어가게 되어 중괄호를 한 단계 덜 들어간다는 장점이 있습니다. 단 guard를 사용할 때는 반드시 else 문이 들어가야 합니다.
- import Foundation : Foundation은 swift의 기본 뼈대가 정리된 프레임워크로, 문자형, 컬렉션과 같은 데이터타입을 포함한 전반적인 내용을 이용하기 위해 반드시 임포트되어 있어야 합니다.
- Set은 STEP 2에서 활용할 예정입니다!
